src/messaging/ExchangeTransport.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file provides implementation of ExchangeTransport class.
+ */
+
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <inttypes.h>
+
+#include <messaging/ExchangeTransport.h>
+#include <messaging/ReliableMessageContext.h>
+#include <messaging/ReliableMessageMgr.h>
+#include <protocols/secure_channel/Constants.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+namespace Messaging {
+
+CHIP_ERROR ExchangeTransport::SendMessage(SecureSessionHandle session, ExchangeTransport::ExchangeInfo & exchangeInfo,
+                                          ReliableMessageContext & reliableMessageContext, bool isReliableTransmission,
+                                          Protocols::Id protocol, uint8_t type, System::PacketBufferHandle message)
+{
+    ReturnErrorCodeIf(!MessagePermitted(protocol.GetProtocolId(), type), CHIP_ERROR_INVALID_ARGUMENT);
+
+    PayloadHeader payloadHeader;
+    payloadHeader.SetExchangeID(exchangeInfo.mExchangeId).SetMessageType(protocol, type).SetInitiator(exchangeInfo.mInitiator);
+
+    // If there is a pending acknowledgment piggyback it on this message.
+    if (reliableMessageContext.HasPeerRequestedAck())
+    {
+        payloadHeader.SetAckId(reliableMessageContext.mPendingPeerAckId);
+
+        // Set AckPending flag to false since current outgoing message is going to serve as the ack on this exchange.
+        reliableMessageContext.SetAckPending(false);
+
+#if !defined(NDEBUG)
+        if (!payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::StandaloneAck))
+        {
+            ChipLogProgress(ExchangeManager, "Piggybacking Ack for MsgId:%08" PRIX32 " with msg",
+                            reliableMessageContext.mPendingPeerAckId);
+        }
+#endif
+    }
+
+    if (!IsTransportReliable() && reliableMessageContext.AutoRequestAck() && mReliableMessageMgr != nullptr &&
```

Comment:
It looks isReliableTransmission is just translated from it is requested, we should also check the low level Transport type. CRMP is not needed for TCP

Reply:
The expectation is that `IsTransportReliable()` method for TCP based exchange transport (e.g. `SessionEstablishmentExchangeTransportTCP`) will return `True`.

Reply:
ApplicationExchangeDispatch should support both TCP and UDP right? if I run chip-shell ping -t <ip address>, the message should also go through ApplicationExchangeDispatch, right?

Reply:
ApplicationExchangeDispatch will be used by IM messages. Those should all be UDP only.

Reply:
Lets have a ticket to follow up here, since all of our examples still support TCP option now

Reply:
https://github.com/project-chip/connectedhomeip/issues/6026